modules/core/src/quaternion.cpp

```diff
@@ -0,0 +1,742 @@
+    qz = z * q1.w + w * q1.z + x * q1.y - y * q1.x;
+    w = qw;
+    x = qx;
+    y = qy;
+    z = qz;
+    return *this;
+}
+
+template <typename T>
+Quat<T>& Quat<T>::operator*=(const T &q1)
+{
+    w *= q1;
+    x *= q1;
+    y *= q1;
+    z *= q1;
+    return *this;
+}
+
+template <typename T>
+inline Quat<T>& Quat<T>::operator/=(const T &a)
+{
+    w /= a;
+    x /= a;
+    y /= a;
+    z /= a;
+    return *this;
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::operator/(const T &a) const
+{
+    return Quat<T>(w / a, x / a, y / a, z / a);
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::operator/(const Quat<T> &q) const
+{
+    return *this * q.inv();
+}
+
+template <typename T>
+inline const T& Quat<T>::operator[](std::size_t n) const
+{
+    switch (n) {
+        case 0:
+            return w;
+            break;
+        case 1:
+            return x;
+            break;
+        case 2:
+            return y;
+            break;
+        case 3:
+            return z;
+            break;
+        default:
+            throw ("subscript exceeds the index range");
+    }
+}
+
+template <typename T>
+inline T& Quat<T>::operator[](std::size_t n)
+{	
+    switch (n) {
+        case 0:
+            return w;
+            break;
+        case 1:
+            return x;
+            break;
+        case 2:
+            return y;
+            break;
+        case 3:
+            return z;
+            break;
+        default:
+            throw ("subscript exceeds the index range");
+    }
+}
+
+template <typename T>
+std::ostream & cv::operator<<(std::ostream &os, const Quat<T> &q)
+{
+    os << "Quat " << cv::Vec<T, 4>{q.w, q.x, q.y, q.z};
+    return os;
+}
+
+template <typename T>
+inline Quat<T> Quat<T>::conjugate() const
+{
+    return Quat<T>(w, -x, -y, -z);
+}
+
+template <typename T>
+inline T Quat<T>::norm() const
+{
+    return std::sqrt(dot(*this));
+}
+
+/*
+template <typename T>
+Quat<T> cv::exp(const Quat<T> &q)
+{
+    return q.exp();
+}
+*/
+
+template <typename T>
+Quat<T> Quat<T>::exp() const
+{
+    cv::Vec<T, 3> v{x, y, z};
+    T normV = std::sqrt(v.dot(v));
+    T k = normV < EPS ? 1 : std::sin(normV) / normV;
+    return std::exp(w) * Quat<T>(std::cos(normV), v[0] * k, v[1] * k, v[2] * k);
+}
+
+/*
+template <typename T>
+Quat<T> cv::log(const Quat<T> &q)
+{
+    return q.log();
+}
+*/		
+
+template <typename T>
+Quat<T> Quat<T>::log() const
+{
+    cv::Vec<T, 3> v{x, y, z};
+    T vNorm = std::sqrt(v.dot(v));
+    T qNorm = norm();
+    T k = vNorm < EPS ? 1 : std::acos(w / qNorm) / vNorm;
+    return Quat<T>(std::log(qNorm), v[0] * k, v[1] * k, v[2] *k);
+}
+
+/*
+template <typename T>
+inline Quat<T> cv::power(const Quat<T> &q1, T x)
+{
+    return q1.power(x);
+}
+*/
+
+template <typename T>
+inline Quat<T> Quat<T>::power(T x1) const
+{
+    T angle = getAngle();
+    cv::Vec<T, 3> axis = getAxis();
+    return std::pow(norm(), x1) * Quat<T>(x1 * angle, axis);
+}
```

Comment:
This function can have an `assumeUnit` flag as well as a bunch of other functions.

Reply:
Ok, thanks. I misunderstood it, and it has been added in latest code. I still have some tests on trigonometric functions left to see the fixed code.